### ISSUE TITLE
termux-setup-storage could still delete files unexpectedly?

### ISSUE THREAD

**Issue by tomty89** [termux/termux-app#738]
So I saw the recent issue and commit that attempted to fix it, and notice that it could still be dangerous if users made a ~/storage a symlink to some directory before using the command, so I did a test:
```
$ mkdir /data/data/com.termux/files/testdir
$ touch /data/data/com.termux/files/testdir/a
$ mkdir /data/data/com.termux/files/testdir/b
$ touch /data/data/com.termux/files/testdir/b/c
$ ln -s /data/data/com.termux/files/testdir $HOME/storage
$ tree /data/data/com.termux/files/testdir
/data/data/com.termux/files/testdir
├── a
└── b
    └── c

1 directory, 2 files
$ termux-setup-storage
$ tree /data/data/com.termux/files/testdir
/data/data/com.termux/files/testdir
└── b
    └── c

1 directory, 1 file
$
```

It isn't as dangerous as I thought, but still you can see that files got deleted unexpectedly.

Tried to write something to do the task myself. Not heavily tested but seems to work. You can take it as a *reference*:
```
import java.io.File;
class recursiveDelete {
    static void deleteNoFollow(File file) {
	if (!file.delete()) {
	    File[] children = file.listFiles();
	    for (File child: children)
		deleteNoFollow(child);
	    file.delete();
	}
    }
    public static void main(String[] args) {
	File blah = new File(args[0]);
        deleteNoFollow(blah);
    }
}
```

P.S. I never get why would someone ever wanted/needed `termux-setup-storage` though. (I always uses the convenient `/sdcard` symlink that the system has for me; probably a bit trickier if one uses a real SD card though.)

**Comment by xeffyr** [comment]
@tomty89 I run commands from your test and got this:
```
$ ls
myfiles
$ mkdir /data/data/com.termux/files/testdir
$ touch /data/data/com.termux/files/testdir/a
$ mkdir /data/data/com.termux/files/testdir/b
$ touch /data/data/com.termux/files/testdir/b/c
$ ln -s /data/data/com.termux/files/testdir $HOME/storage
$ tree /data/data/com.termux/files/testdir
/data/data/com.termux/files/testdir
├── a
└── b
    └── c

1 directory, 2 files
$ termux-setup-storage
$ tree /data/data/com.termux/files/testdir
/data/data/com.termux/files/testdir
├── a
└── b
    └── c

1 directory, 2 files
$ ls -ld storage
lrwxrwxrwx 1 u0_a309 u0_a309 35 Jun 25 17:59 storage -> /data/data/com.termux/files/testdir
```

You have file `/data/data/com.termux/files/testdir/a` disappeared but I not, thats strange. But at least it don't remove everything from internal storage.

**Comment by tomty89** [comment]
The thing is I get the logic of the behavior though (I think).

First of all `deleteFolder` doesn't check whether the argument passed to it is a symlink.

Then it goes check its children:
```
                if (child.getCanonicalFile().equals(child.getAbsoluteFile())) {
                    deleteFolder(child);
                } else {
                    child.delete();
                }
```
And this is what happened:
```
/data/data/com.termux/files/testdir/a
/data/data/com.termux/files/home/storage/a
/data/data/com.termux/files/testdir/b
/data/data/com.termux/files/home/storage/b
```
But then even when the tests failed, it attempts to delete the child (the `else` clause). `b` didn't get deleted only because it's non-empty, otherwise it gets deleted as well.

**Comment by tomty89** [comment]
By the way the idea of my example code is, always attempt to delete a file before traversing into (or listing, actually), and only do that if the deletiom failed. It should work unless something like this happened: it bumps into a symlink that can't be deleted, yet it can find files via it and delete them.

In any case I still oppose the idea of `termux-setup-storage` though. At least the part that it attempts to perform any form of deletion (why do we expect no one would put files under ~/storage before realizing it could get deleted by the app?). Not to mention the symlinks get created is IMHO a bit pointless. Can't we just have it print out the available storage paths?

**Comment by xeffyr** [comment]
@tomty89 I have switched to OpenJDK 8 instead of Oracle's JDK 9 and was able to reproduce the problem with deleting files as in your test.

Fix for this issue is available now.

**Comment by xeffyr** [comment]
@tomty89 I have switched to OpenJDK 8 instead of Oracle's JDK 9 and was able to reproduce the problem with deleting files as in your test.

Fix for this issue is available now.

P.S. Building android application with JDK-9 seems a bad idea.

**Comment by xeffyr** [comment]
This issue fixed in v0.64 of Termux app so it should be closed.
